lib/applicants/resolvers.ts

```diff
@@ -8,14 +8,146 @@ import {
+ *
+ * Sorting:
+ * - order: array of tuples of the field being sorted and the order. Default [['firstName', 'asc'], ['lastName', 'asc']]
+ * @returns All RCD applicants that match the filter(s).
  */
-export const applicants: Resolver = async (_parent, _args, { prisma }) => {
-  const applicants = await prisma.applicant.findMany();
-  return applicants;
+export const applicants: Resolver = async (_parent, { filter }, { prisma }) => {
+  // Create default filter
+  let where = {};
+
+  if (filter) {
+    const {
+      permitStatus = undefined,
+      userStatus = undefined,
+      expiryDateRangeFrom = undefined,
+      expiryDateRangeTo = undefined,
+      search = undefined,
+    } = filter;
+
+    let expiryDateUpperBound,
+      expiryDateLowerBound,
+      userIDSearch,
+      nameFilters,
+      firstSearch,
+      middleSearch,
+      lastSearch;
+
+    // Parse search input for id or name
+    if (parseInt(search)) {
+      userIDSearch = parseInt(search);
+    } else if (search) {
+      // Split search assign to first, middle or last name
+      // If one doesn't exist assign to the previous input to allow for global search across names
+      [firstSearch, middleSearch, lastSearch] = search?.split(' ');
+      middleSearch = middleSearch || firstSearch;
+      lastSearch = lastSearch || middleSearch;
+
+      nameFilters = [
+        { firstName: { contains: firstSearch, mode: 'insensitive' } },
+        { middleName: { contains: middleSearch, mode: 'insensitive' } },
+        { lastName: { contains: lastSearch, mode: 'insensitive' } },
+      ];
+    }
+
+    const TODAY = new Date();
+
+    // Permit status filter depends on expiry date
+    switch (permitStatus) {
+      case PermitStatus.Valid:
+        expiryDateLowerBound = TODAY;
+        break;
+      case PermitStatus.Expired:
+        expiryDateUpperBound = TODAY;
+        break;
+      case PermitStatus.ExpiringThirty:
+        expiryDateLowerBound = TODAY;
+        expiryDateUpperBound = DateUtils.addMonths(TODAY, 1);
+        break;
+    }
+
+    const containsPermitFilter = !!(permitStatus || expiryDateRangeFrom || expiryDateRangeTo);
+
+    // Permit status filter and expiry date range both use the permit expiry date.
+    // For this reason we need to filter on expiry date twice to return an accurate result.
+    const permitFilter = containsPermitFilter
+      ? {
+          some: {
+            AND: [
+              {
+                expiryDate: {
+                  gte: expiryDateLowerBound?.toISOString(),
+                  lte: expiryDateUpperBound?.toISOString(),
+                },
+              },
+              {
+                expiryDate: {
+                  gte: expiryDateRangeFrom?.toISOString(),
+                  lte: expiryDateRangeTo?.toISOString(),
+                },
+              },
+            ],
+          },
+        }
+      : undefined;
+
+    // Update default filter since there were filter arguments
+    where = {
```

Comment:
I think that another filter is required here to only return applicants that have a "recent permit", as in they have at least one permit associated with them. Not 100% this is even a valid scenario, but it does seem reasonable that an applicant could be created before a permit is associated with them. I got some errors in `pages/admin/permit-holders` due to getting null results for the `recentPermit` and `status` fields.

Reply:
Good point! I double checked with Oustan abt this and it turns out every applicant entry must have at least one associated permit since to become an 'applicant' they'd need to have their application approved which gives them a permit in the process. I changed the seed data to better reflect this

As a sanity check tho I also added a clause to make sure every applicant has at least one permit and changed the gql schema too in my latest commit